### PR TITLE
Introduce notations

### PR DESCRIPTION
**Pull request by etiennecallies**
Not implemented yet:
 - [x] coq export with notations
 - [ ] latex notations definition
 - [ ] auto-prover with notations

### CODE REVIEW

File: is_sequent_provable.ml [hunk 1]
```diff
@@ -1,10 +1,22 @@
...
+    check_sequent_provability sequent_with_notations.sequent || begin
+        let cyclic_notations, sorted_acyclic_notations = Sequent_with_notations.split_cyclic_acyclic sequent_with_notations in
+        if List.length cyclic_notations > 0
+        then true
+        else if List.length sorted_acyclic_notations = 0
+            then false
+            else
+                let sequent = Sequent_with_notations.replace_all_notations_in_sequent sequent_with_notations.sequent sorted_acyclic_notations in
+                check_sequent_provability sequent
+    end;;
```
Comment (olaure01) [inline is_sequent_provable.ml:17]:
```suggestion
    let cyclic_notations, sorted_acyclic_notations = Sequent_with_notations.split_cyclic_acyclic sequent_with_notations in
    if List.length cyclic_notations > 0
    then true
    else let sequent = Sequent_with_notations.replace_all_notations_in_sequent sequent_with_notations.sequent sorted_acyclic_notations in
         check_sequent_provability sequent
```

File: sequent_with_notations.ml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+exception Json_exception of string
+
+let get_key d k =
+    let value =
+        try Yojson.Basic.Util.member k d
+        with Yojson.Basic.Util.Type_error (_, _) -> raise (Json_exception ("request body must be a json object"))
+    in
+    if value = `Null
+    then raise (Json_exception ("required argument '" ^ k ^ "' is missing"))
+    else value
```
Comment (olaure01) [inline sequent_with_notations.ml:15]: Same code in `proof_with_notations.ml`, move to `notations.ml`?

File: sequent_with_notations.ml [hunk 3]
```diff
@@ -0,0 +1,96 @@
+open Sequent
+open Notations
+
+type proof_with_notations = {sequent: sequent; notations: notations}
```
Comment (olaure01) [inline sequent_with_notations.ml:4]:
```suggestion
type sequent_with_notations = {sequent: sequent; notations: notations}
```

File: sequent.ml [hunk 4]
```diff
@@ -72,6 +72,20 @@ let sort sequent =
...
 
 
+let rec replace_in_formula f1 f2 = function
+    | Tensor (e1, e2) -> Tensor (replace_in_formula f1 f2 e1, replace_in_formula f1 f2 e2)
+    | Par (e1, e2) -> Par (replace_in_formula f1 f2 e1, replace_in_formula f1 f2 e2)
+    | With (e1, e2) -> With (replace_in_formula f1 f2 e1, replace_in_formula f1 f2 e2)
+    | Plus (e1, e2) -> Plus (replace_in_formula f1 f2 e1, replace_in_formula f1 f2 e2)
+    | Ofcourse e -> Ofcourse (replace_in_formula f1 f2 e)
+    | Whynot e -> Whynot (replace_in_formula f1 f2 e)
+    | f when f = f1 -> f2
```
Comment (olaure01) [inline sequent.ml:82]: Either define substitution on a litteral: `replace_in_formula  (s : string) (f : formula) : formula` (which substitutes both `Litt s` with `f` and `Dual s` with `dual f`)
or the `| f when f = f1 -> f2` matching should probably be on top position.

File: notations.ml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+open Sequent
+
+type notations = (string * formula) list;;
```
Comment (olaure01) [inline notations.ml:3]: probably `raw_formula` for LaTeX export

File: auto_prove_sequent.ml [hunk 6]
```diff
@@ -1,13 +1,24 @@
...
+    try Some (Proof.auto_weak sequent_with_notations.sequent), true
     with Proof.AutoWeakNotApplicable ->
-        Focused_proof.prove_sequent sequent;;
+        let result = try_auto_prover_on_sequent sequent_with_notations.sequent in
+        match result with
+        | Some proof, _ -> result
+        | None, _ ->
+            let cyclic_notations, acyclic_notations = Sequent_with_notations.split_cyclic_acyclic sequent_with_notations in
+            if List.length cyclic_notations + List.length acyclic_notations > 0
+            then None, true
```
Comment (olaure01) [inline auto_prove_sequent.ml:16]: explain in documentation that orange ⊢ means either time out or cyclic notations

File: notations.ml (hunk 5 above)
Comment (etiennecallies) [reply notations.ml]: I just did it thanks.

File: is_sequent_provable.ml (hunk 1 above)
Comment (etiennecallies) [reply is_sequent_provable.ml]: You're right, it's not worth it to bring that confusion in the code.

Comment (etiennecallies) [reply is_sequent_provable.ml]: Done

File: sequent_with_notations.ml (hunk 2 above)
Comment (etiennecallies) [reply sequent_with_notations.ml:15]: I agree request json handling deserves a refactoring, that I'll do later on.

File: sequent.ml (hunk 4 above)
Comment (etiennecallies) [reply sequent.ml]: I opted for first option, thanks!

File: auto_prove_sequent.ml (hunk 6 above)
Comment (etiennecallies) [reply auto_prove_sequent.ml:16]: Done. I will precise that it fails only with recursive notations, when the autoprover for non-recursive notations will be implemented.

**Comment by etiennecallies** [comment]
Thanks @olaure01 👍🏻

**Pull request merged**